File: somestring/hw8/frontend/src/app/apiFunctions/apiFunctions.ts
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { latestPrice } from '../interfaces/latestPrice';
import { companyDescription } from '../interfaces/companyDescription';
import { historicalData } from '../interfaces/historicalData'
import { Autocomplete } from '../interfaces/autocomplete';
import { companyNews } from '../interfaces/companyNews';
import { companySocialSentiment } from '../interfaces/companySocialSentiment';
import { recommendationTrends } from '../interfaces/recommendationTrends';
import { companyEarnings } from '../interfaces/companyEarnings';

const HOST = 'http://localhost:3000';

@Injectable({
  providedIn: 'root',
})
export class APIService {
  private autoComplete = HOST + '/autocomplete';
  private latestPrice = HOST + '/getLatestPrice';
  private companyDescription = HOST + '/getCompanyDescription';
  private historicalData = HOST + '/getHistoricalData';
  private companyNews = HOST + '/getCompanyNews';
  private companySocialSentiment = HOST + '/getCompanySocialSentiment';
  private companyRecommendationTrends = HOST + '/getCompanyRecommendationTrends';
  private companyEarnings = HOST + '/getCompanyEarnings';
  private companyPeers = HOST + '/getCompanyPeers';



  constructor(private http: HttpClient) {}

  fetchAutocomplete(ticker: string): Observable<Autocomplete[]> {
    const url = `${this.autoComplete}?query=${ticker}`;
    return this.http.get<Autocomplete[]>(url);
  }

  fetchCompanyDescription(ticker: string): Observable<companyDescription> {
    const url = `${this.companyDescription}?ticker=${ticker}`;
    return this.http.get<companyDescription>(url);
  }

  fetchLatestPrice(ticker: string): Observable<latestPrice> {
    const url = `${this.latestPrice}?ticker=${ticker}`;
    return this.http.get<latestPrice>(url);
  }
  
  fetchHistoricalData(ticker, timeInterval, fromTimestamp, toTimestamp): Observable<historicalData> {
    const url = `${this.historicalData}?ticker=${ticker}&timeInterval=${timeInterval}&fromTimestamp=${fromTimestamp}&toTimestamp=${toTimestamp}`;
    return this.http.get<historicalData>(url);
  }

  fetchCompanyNews(ticker, fromDate, toDate): Observable<companyNews[]> {
    const url = `${this.companyNews}?ticker=${ticker}&fromDate=${fromDate}&toDate=${toDate}`;    
    return this.http.get<companyNews[]>(url);
  }

  fetchCompanySocialSentiment(ticker): Observable<companySocialSentiment> {
    const url = `${this.companySocialSentiment}?ticker=${ticker}`;    
    return this.http.get<companySocialSentiment>(url);
  }

  fetchCompanyRecommendationTrends(ticker): Observable<recommendationTrends[]> {
    const url = `${this.companyRecommendationTrends}?ticker=${ticker}`;    
    return this.http.get<recommendationTrends[]>(url);
  }

  fetchCompanyEarnings(ticker): Observable<companyEarnings[]> {
    const url = `${this.companyEarnings}?ticker=${ticker}`;    
    return this.http.get<companyEarnings[]>(url);
  }

  fetchCompanyPeers(ticker): Observable<string[]> {
    const url = `${this.companyPeers}?ticker=${ticker}`;    
    return this.http.get<string[]>(url);
  }
  
}

File: somestring/hw9/csci571hw9/csci571hw9/APIFunctions.swift
//
//  APIFunctions.swift
//  csci571hw9
//
//  Created by Jerry Lee on 4/30/22.
//

import Foundation
import Alamofire
import simd

class APIFunctions: ObservableObject {
    var ticker: String = "";
    
    
    var isLoading: Bool = true;
    var autoCompleteURL: String = "https://directed-sonar-346104.wl.r.appspot.com/autocomplete";
    var latestPriceURL: String = "https://directed-sonar-346104.wl.r.appspot.com/getLatestPrice";
    var companyDescriptionURL: String = "https://directed-sonar-346104.wl.r.appspot.com/getCompanyDescription";
    var companyNewsURL: String = "https://directed-sonar-346104.wl.r.appspot.com/getCompanyNews";
    var companySocialSentimentURL: String = "https://directed-sonar-346104.wl.r.appspot.com/getCompanySocialSentiment";
    var companyPeersURL: String = "https://directed-sonar-346104.wl.r.appspot.com/getCompanyPeers";
    
    @Published var autoComplete: [Autocomplete] = []
    @Published var latestPrice: LatestPrice?
    @Published var companyDescription: CompanyDescription?
    @Published var companyNews: [CompanyNews]?
    @Published var companySocialSentiment: filteredSocialSentiment?
    @Published var companyPeers: [String]?
    
    func fetchAutoComplete(ticker: String) {
        guard let url = URL(string: "\(autoCompleteURL)?query=\(ticker)")
        else {
            print("URL doesn't exist")
            return
        }
        AF.request(url).responseDecodable(of: [Autocomplete].self ) { (response) in
            guard let res = response.value else {return}
            let filteredRes = res.filter{ val in return !val.symbol.contains(".")}
            self.autoComplete = filteredRes
        }
    }
    
    func fetchLatestPrice(ticker: String) {
        guard let url = URL(string: "\(latestPriceURL)?ticker=\(ticker)")
        else {
            print("URL doesn't exist")
            return
        }
        AF.request(url).responseDecodable(of: LatestPrice.self ) { (response) in
            guard let res = response.value else {return}
            self.latestPrice = res
        }
    }
    
    func fetchCompanyDescription(ticker: String) {
        guard let url = URL(string: "\(companyDescriptionURL)?ticker=\(ticker)")
        else {
            print("URL doesn't exist")
            return
        }
        AF.request(url).responseDecodable(of: CompanyDescription.self ) { (response) in
            guard let res = response.value else {return}
            self.companyDescription = res
        }
    }
    
    func fetchCompanyNews(ticker: String) {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let toDate = Date()
        let fromDate = toDate.addingTimeInterval(-86400*7)
        let toDateFormatted = formatter.string(from: toDate)
        let fromDateFormatted = formatter.string(from: fromDate)
        
        guard let url = URL(string: "\(companyNewsURL)?ticker=\(ticker)&fromDate=\(fromDateFormatted)&toDate=\(toDateFormatted)")
        else {
            print("URL doesn't exist")
            return
        }
        AF.request(url).responseDecodable(of: [CompanyNews].self ) { (response) in
            guard let res = response.value else {return}
            let filteredRes = res.filter{val in return !(val.image == "")}
            self.companyNews = filteredRes.sorted(by: {$0.datetime > $1.datetime})
            self.isLoading = false
        }
    }
    
    func fetchCompanySocialSentiment(ticker: String) {
        guard let url = URL(string: "\(companySocialSentimentURL)?ticker=\(ticker)")
        else {
            print("URL doesn't exist")
            return
        }
        AF.request(url).responseDecodable(of: CompanySocialSentiment.self ) { (response) in
            guard let res = response.value else {return}
            let filteredReddit = res.reddit.reduce(Reddit(mention: 0, positiveMention: 0, negativeMention: 0), { x, y in
                Reddit(mention: x.mention+y.mention, positiveMention: x.positiveMention+y.positiveMention, negativeMention: x.negativeMention+y.negativeMention)
            })
            let filteredTwitter = res.twitter.reduce(Twitter(mention: 0, positiveMention: 0, negativeMention: 0), { x, y in
                Twitter(mention: x.mention+y.mention, positiveMention: x.positiveMention+y.positiveMention, negativeMention: x.negativeMention+y.negativeMention)
            })
            self.companySocialSentiment = filteredSocialSentiment(symbol: res.symbol, reddit: filteredReddit, twitter: filteredTwitter)
        }
    }
    
    func fetchCompanyPeers(ticker: String) {
        guard let url = URL(string: "\(companyPeersURL)?ticker=\(ticker)")
        else {
            print("URL doesn't exist")
            return
        }
        AF.request(url).responseDecodable(of: [String].self ) { (response) in
            guard let res = response.value else {return}
            let filteredRes = res.filter{ val in return !val.contains(".")}
            self.companyPeers = filteredRes
        }
    }
}
